text/0000-output-options.md

```diff
@@ -0,0 +1,135 @@
+`-o` option for *some* of `--emit` targets, even though it emits a warning about filename being
+ignored. The document aims to propose rules governing interactions of the three options.
+
+# Detailed design
+
+## `--out-dir`
+
+`--out-dir` is never ignored or adjusted, even in presence of other options that would usually
+conflict with `--out-dir`. If the option is provided and the directory does not exist yet, the
+compiler should create it. Defaults to `.`, otherwise known as current working directory or `$PWD`.
+Value of this option from now on will be referred to as `[out-dir]`.
+
+## `--emit` and `--crate-type`
+
+The two options influence what and how many files the compiler has to write to the filesystem. The
+common case compiler having to write one file only (e.g. `--emit` and `--crate-type` only have one
+value each). In this case the path to output file shall be built using these templates:
+
+* `[out-dir]/[filename][.extension]` if [filename] was [inferred][inferred].
+* `[out-dir]/[filename]` otherwise;
+
+[inferred]: #-o-is-not-specified
+
+Otherwise, when there’s multiple files to output, each output file will be written to path built
+using `[out-dir]/[filename][.extension]` template.
+
+What `[filename]` resolves to is specified in [section about `-o`](#-o).
+
+The `[.extension]` is file extension specific to the emitted target. This is a list of extensions
+for each currently supported target:
+
+* asm – `.s`;
+* llvm-bc – `.bc`;
+* llvm-ir – `.ll`;
+* obj – `.o`;
+* dep-info – `.d`;
+* link – depends on `crate-type`:
+  * bin – no extension or `.exe`;
+  * lib – `.rlib`;
+  * rlib – `.rlib`;
+  * dylib – `.so`, `.dll` or `.dylib`;
+  * staticlib – `.a`.
+
+`link` target should prepend prefix `lib` to `[filename]` for all `crate-type`s except `bin`:
+
+    $ rustc foo.rs -o foo --crate-type=staticlib
+    # Output: libfoo.a
+
+## `-o`
+
+The value of `-o` option, unlike `--out-dir` may be adapted by `rustc` in case it conflicts with
+other options. Ignoring the option is strongly discouraged.
+
+Were the value be ignored or adapted, the warning detailing actions taken should be emitted.
+
+### Interactions with `--out-dir`
+
+In absence of `--out-dir` option, `-o` value may contain directory path components:
+
+    $ rustc foo.rs -o bar/foo # A-OK
+
+For this particular command the compiler behaves as if it had gotten options `--out-dir=bar/` and
+`-o foo`. More generally, the compiler interprets the option by splitting value into
+[`filename`][filename] to be used as `[filename]` and [`dirname`][dirname] to be used as
+`[out-dir]`.
+
+[filename]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#tymethod.filename
+[dirname]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#tymethod.dirname
+
+If both `--out-dir` and `-o` are provided, the `-o` value is adjusted so it only contains a
+`filename`, which is used as `[filename]`:
```

Comment:
I just personally found the suggested approach most intuitive out of all possible ways to handle this from the POV of human user (me).

I’ll add this to the list of alternatives, since it is another good approach to handling the conflicts.